### PR TITLE
Zhengli0817 patch 1 - update topsort.py

### PR DESCRIPTION
**Pull request by zhengli0817**
The original algo in topsort.py is wrong because it is using BFS, it should use DFS instead.

A counter-example is provided below:

depGraph = {
"a" : [ "b" ],
"b" : [ "c" ],
"c" : [ 'e'],
'e' : [ 'g' ],
"d" : [ ],
"f" : ["e" , "d"],
"g" : [ ]
}

given = [ "b", "c", "a", "d", "e", "f", "g" ]

The output of ret_dep_graph() is: ['e', 'g', 'c', 'b', 'a', 'd', 'f']

Clearly, 'g' has to be done before 'e'.

### CODE REVIEW

File: algorithms/sort/topsort.py [hunk 1]
```diff
@@ -1,60 +1,79 @@
...
+    return order
 
-    "a": ["b"],
-    "b": ["c"],
-    "c": ['e'],
-    'e': [],
-    "d": [],
-    "f": ["e", "d"]
-}
+class TestSuite(unittest.TestCase):
```
Comment (goswami-rahul) [inline algorithms/sort/topsort.py]: move these tests to the `/tests/graph/` folder please.

File: algorithms/sort/topsort.py [hunk 2]
```diff
@@ -1,60 +1,79 @@
...
+                        "f" : ["e" , "d"],
+                        "g" : [ ]
+                    }
+        
+    def test_order(self):
+        res = topological_sort_recursive(self.depGraph)
+        #print(res)
+        self.assertTrue(res.index('g')<res.index('e'))
+        res = topological_sort_recursive(self.depGraph)
+        self.assertTrue(res.index('g')<res.index('e'))
```
Comment (goswami-rahul) [inline algorithms/sort/topsort.py]: put spaces before and after ` < ` symbol.

File: algorithms/sort/topsort.py [hunk 3]
```diff
@@ -1,60 +1,79 @@
...
+def topological_sort_recursive(graph):
+    order, enter, state = [], set(graph), {}
+    
+    def dfs(node):
+        state[node] = GRAY
+        #print(node)
+        for k in graph.get(node, ()):
+            sk = state.get(k, None)
+            if sk == GRAY: raise ValueError("cycle")
+            if sk == BLACK: continue
```
Comment (goswami-rahul) [inline algorithms/sort/topsort.py]: seperate these `if` statements in 2 lines rather than using a one liner.

**Comment by zhengli0817** [comment]
@goswami-rahul I made the changes according to your comments, please review

File: algorithms/sort/__init__.py [hunk 4]
```diff
@@ -6,3 +6,4 @@
 from .merge_sort import *
 from .quick_sort import *
 from .selection_sort import *
+from .topsort import *
```
Comment (keon) [inline algorithms/sort/__init__.py]: please change this to `top_sort` for uniformity.

File: algorithms/sort/topsort.py [hunk 5]
```diff
@@ -1,60 +1,60 @@
...
+                continue
+            enter.discard(k)
+            dfs(k)
+        order.append(node)
+        state[node] = BLACK
+        
+    while enter: dfs(enter.pop())
+    return order
+
+def topological_sort(graph):
```
Comment (keon) [inline algorithms/sort/topsort.py]: i think we could just use `top_sort` as it is well received as the short term for topological sort.

File: algorithms/sort/topsort.py [hunk 6]
```diff
@@ -1,60 +1,60 @@
...
-            continue
-        visited.add(pac)
-        # out.append(pac)
-        if pac in depGraph:
-            # find all children
-            for child in depGraph[pac]:
-                if child in visited:
+GRAY, BLACK = 0, 1
+
+def topological_sort_recursive(graph):
```
Comment (keon) [inline algorithms/sort/topsort.py]: could you add a little explanation (or link to the explanation)
and time/space complexity?

**Comment by zhengli0817** [comment]
@goswami-rahul I made the changes again according to your comments, please review

**Comment by keon** [comment]
@zhengli0817 merged! thanks :)

**Pull request merged**